daemon/src/routes_taker.rs

```diff
@@ -159,27 +161,55 @@ pub async fn post_cfd_action(
 }
 
 #[derive(Debug, Clone, Deserialize)]
-pub struct WalletReinitialiseRequest {
-    pub seed_words: String,
+pub struct WalletRestoreRequest {
+    pub mnemonic: String,
 }
 
-#[rocket::post("/wallet/reinitialise", data = "<wallet_reinitialise_request>")]
-pub async fn post_wallet_reinitialise(
-    wallet_reinitialise_request: Json<WalletReinitialiseRequest>,
+#[rocket::post("/wallet/mnemonic", data = "<wallet_restore_request>")]
+pub async fn post_wallet_mnemonic(
+    wallet_restore_request: Json<WalletRestoreRequest>,
     taker: &State<Taker>,
 ) -> Result<status::Accepted<()>, HttpApiProblem> {
-    taker
-        .reinitialise_wallet(&wallet_reinitialise_request.seed_words)
-        .await
-        .map_err(|e| {
-            HttpApiProblem::new(StatusCode::INTERNAL_SERVER_ERROR)
-                .title("Wallet recover request failed")
-                .detail(e.to_string())
-        })?;
+    let mnemonic = Mnemonic::from_str(&wallet_restore_request.mnemonic).map_err(|e| {
+        HttpApiProblem::new(StatusCode::INTERNAL_SERVER_ERROR)
+            .title("Provided mnemonic is invalid")
+            .detail(e.to_string())
+    })?;
+
+    taker.restore_wallet(mnemonic).await.map_err(|e| {
+        HttpApiProblem::new(StatusCode::INTERNAL_SERVER_ERROR)
+            .title("Wallet restore request failed")
+            .detail(e.to_string())
+    })?;
 
     Ok(status::Accepted(None))
 }
 
+#[rocket::get("/wallet/mnemonic")]
+pub async fn get_wallet_mnemonic(
+    taker: &State<Taker>,
+) -> Result<status::Accepted<String>, HttpApiProblem> {
+    let mnemonic = taker.backup_wallet().await.map_err(|e| {
+        HttpApiProblem::new(StatusCode::INTERNAL_SERVER_ERROR)
+            .title("Wallet recover request failed")
+            .detail(e.to_string())
+    })?;
+
+    Ok(status::Accepted(Option::from(mnemonic.to_string())))
+}
+
+#[rocket::post("/mnemonic")]
+pub async fn post_mnemonic(
+    taker: &State<Taker>,
+) -> Result<status::Accepted<String>, HttpApiProblem> {
+    let mnemonic = taker.generate_mnenomic().map_err(|e| {
+        HttpApiProblem::new(StatusCode::INTERNAL_SERVER_ERROR)
+            .title("Generate new mnemonic request failed")
+            .detail(e.to_string())
+    })?;
+    Ok(status::Accepted(Option::from(mnemonic.to_string())))
+}
```

Comment:
This interface seems to be unnecessary chatty:

The two functionalities we need are:

1. Restore a wallet from a mnemonic
2. Re-initialize the wallet with a new one

Currently, for (1), we would need to do:

```
POST /wallet/mnemonic
```

And for (2), we would need to do:

```
POST /mnemonic
POST /wallet/mnemonic
```

I don't think we need to split (2) into two requests. What do you think of the following interface:

1. `POST /wallet/restore`

Takes seedwords in the body and re-initializes the wallet with the given ones.

2. `POST /wallet/reinitialize`

Has no request body, generates random seedwords, re-initializes the wallet and returns the seedwords used in the response body.

A design note: Try to design your API around the functionality you want to offer to the client, not so much in the individual steps it takes to get that functionality done. In this case, we have two different things the user may want to achieve. Creating **one** API endpoint per usecase allows us to better capture the users intent and execute the appropriate action. The resulting API is also likely to be more stable because it will only change with those specific requirements.

Reply:
> The two functionalities we need are:
> 
>     1. Restore a wallet from a mnemonic
> 
>     2. Re-initialize the wallet with a new one

We also need the backup functionality in case the user wants to store the mnemonic somewhere for later use:
 
3. Backup a wallet by showing the user the mnemonic 

> Takes seedwords in the body and re-initializes the wallet with the given ones.
> 
>     POST /wallet/reinitialize

yep makes sense to package mnemonic generation and restore together

Reply:
> We also need the backup functionality in case the user wants to store the mnemonic somewhere for later use:
> 
>     3. Backup a wallet by showing the user the mnemonic

True! But that is just covered by `GET /wallet/mnemonic` which I would probably just change to `GET /wallet/seedwords`.